### DISCUSSION TITLE
Revisiting the rendering of commit messages

### DISCUSSION THREAD

**Discussion by Roshick** [renovatebot/renovate#16870] [Suggest an Idea]
Greetings!

In the context of [my issue here](https://github.com/renovatebot/renovate/issues/16839) I came across [this commit](https://github.com/renovatebot/renovate/commit/3b0c2afe8a373b93b810186ac2c021dc5a0342df) which came quite close to solving my issue but had to be reverted due to some bugs. After trying to integrate some hacky solution for my problem, I decided to revisit the commit and implement the idea myself to see where it went wrong because I feel like this is in the context of the project the most elegant solution right now.

In my (fairly limited) view I think I have found the issues in the commit and would like to propose solutions and have a discussion about this topic to hopefully finally get this imho needed refactoring done.

So why do I think the last approach failed? The main reason is [this part](https://github.com/renovatebot/renovate/blob/ab7ddb5e3c5c3b402acd7c3679d4e415f8092dde/lib/config/options/index.ts#L1597), as rendering the subject of the commitMessage [here](https://github.com/Roshick/renovate/blob/8dce65eb6f5859a1b768dcc1efcd7d502425e6b5/lib/workers/repository/updates/generate.ts#L185) while the template still contains the '{{{commitMessagePrefix}}}' would cause the subject to also contain the prefix, creating it twice in the toString method (causing the "\<prefix\>: \<prefix\> \<subject\>" bug).

The second problem is an inconsistency in creating the prefix between the toString method of the CommitMessages (for e.g. onboard branches) and the rendering in the generateBranchConfig function. Currently every commit message being created by the factory will have a prefix separated by its subject by a ':' due to [this function](https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/model/commit-message.ts#L19-L29), while the generateBranchConfig function will simply join them by a whitespace due to the template definition (causing the message and pr to look like this "\<prefix\>: \<subject\>" instead of this "\<prefix\> \<subject\>").

In order to solve this issue I would suggest to consistently create those messages between different flows (e.g. onboarding and updates), which might break in some cases for now (as messages would have to be changed), but favouring the update messages this should not create too much trouble. [My branch ](https://github.com/Roshick/renovate/tree/feat/16839-conventional-commits-uppercase) has a scuffed but working solution if you want to take a look at it.

**Comment by rarkins** [comment]
Do you know in which cases it "might break"? Do you mean they would break "correctly" though, i.e. the new message would be correct?

**Reply by Roshick** [reply]
As far as I understand the code right now every commit created in flows that already use the factory separate prefix from subject via a ':' i.e. every onboarding pr would be created with the commit message and pr title looking like "\<prefix\>: \<subject\>" whereas the update simply creates a "\<prefix\> \<subject\>" commit message and pr title. If we change it to one or the other it will "break correctly" (in the sense that prs that have been declined might be created again) for the other and I would strongly prefer the commit messages and title to be "\<prefix\> \<subject\>". We could also create another toString method or add some parameter that controls whether the ':' should be added or not.

The first problem is probably easy to fix as the prefix should just be removed from the commitMessage template (given no sideeffects by removing it), but I haven't found the definition of those templates yet.

edit: [found it](https://github.com/renovatebot/renovate/blob/ab7ddb5e3c5c3b402acd7c3679d4e415f8092dde/lib/config/options/index.ts#L1597). Removing it fixed the problem of having two prefixes.

**Reply by Roshick** [reply]
I had some more thoughts about the commitMessage template and why it's troubling in this context i.e. there are situations in which my solution will break.

As soon as commitMessagePrefix is set it will be set as such in the CustomCommitMessage as separate part to the subject while the subject is set to the commitMessage template. What the factory commit message will render is "\<prefix\> {{commitMessage}}", therefore fixing the issue by removing the commitMessagePrefix in the default template. Unfortunately this also means that as soon as the commitMessage template contains the commitMessagePrefix (by deviating from the default), the issue will pop up again.

Now I've seen that the commitMessage template parameter is marked as deprecated, are there plans to drop it altogether anytime soon? It makes a proper refactoring of this part of the code very difficult without side effects.

**Reply by rarkins** [reply]
Do I understand correctly, you mean that if we were to fully remove the capability to customize `commitMessage` (instead enforcing our default structure of subelements of commitMessage) then it would make the refactoring easier? For record it is currently:

```
      '{{{commitMessagePrefix}}} {{{commitMessageAction}}} {{{commitMessageTopic}}} {{{commitMessageExtra}}} {{{commitMessageSuffix}}}',
```

**Reply by Roshick** [reply]
This depends on the ultimate goal. Should every commit message be rendered equally (but in different flavours such as Semantic, Custom...) across all flows via the CommitMessageFactory (which makes totally sense to me)? Then for sure. Otherwise you will have two conflicting concepts: 1. Rendering the commit message freely based on the template alone and 2. Rendering the commit message in a (currently partially) static structure in the CommitMessage classes. What we currently have to do is put the free template into the static structure and hope for the best.

I will take another look at it when I have some time to spare and see If I can find a solution that will currently not break, but it's certainly not easy.

Freedom is nice but always comes with complexity, therefore: KISS.

**Reply by rarkins** [reply]
I'm ok with limiting freedom if it makes things easier. Right now, our commit message factory seems too complex to me
